doc/design/feed_op_and_fectch_op.md

```diff
@@ -0,0 +1,120 @@
+# FeedOp and FetchOp Design Doc
+
+### Motivation
+
+Python programer needs an interface to feed the data to PaddlePaddle, run the model, and fetch the result from it. Since PaddlePaddle runtime only goes through a graph of ops, we need to design corresponding Ops and add them to the graph.
+
+### Challenge
+
+1. During the runtime of a particular Op, it only knows which `Variable` to be read from and written to. It doesn't have a direct access to python object.
+2. I/O involves copying data between python object and C++ object. In the current design, python only passes `ProgramDesc` to C++.
+
+### Solution
+
+To solve the first challenge, we add two **global** `Variable`s, `feed_result` and `fetch_result`. The operator can read/write these variables and later on it can be exchanged with python interface. The content of each is a `map<string, LoDTensor>`, where `string` is the `variable.name` and `LoDTensor` is the actual data. Be aware that the actually data is always on CPU.
+
+To solve the second challenge, we add two python interface `feed_value` and `fetch_value`. 
+
+```Python
+def feed_value(variable, np_variable):
+    """Overwrite feed_result[variable.name] with a numpy.array
```

Comment:
`feed_result` and `fetch_value` will be serialized and passed from Python to C++, where are they serialized into? (i.e., are they part of `ProgramDesc`?)

Reply:
Good point. This part must be redesigned.

Reply:
I have an offline discussion with @helinwang.
- Each paddle trainer will have a global Scope and two global Variables(maybe static variable in C++), `feed_result` and `fetch_result`. Python can not create C++ Variable. 
- What Feed Operator does is to take LodTensors from global Variables and copy to its output Variable.
I think that:
- In distributed training, the training data is saved in a distributed file system.  C++ feed_value method will load data from file. And this method must be called before Executor::Run.
- In local machine training, we can exposed feed_value to Python, and numpy array will be set into Global Varibale

Reply:
> In distributed training, the training data is saved in a distributed file system. C++ feed_value method will load data from file.

`feed_value` is only for the argument `feed_dict` in `session.eval`. Which will come from the network. It will never come from the disk. We will have OP that reads data from the disk, but that's is irrelevant with `feed_value`.

> In local machine training, we can exposed feed_value to Python, and numpy array will be set into Global Varibale.

Everything needs to be serialized, "numpy array will be set into Global Varibale" means the Python code is involved in the runtime (which conflicts with the current design).

I think the compile-time_runtime separation is very important. Let's have more discussion if you have other ideas.

Reply:
@helinwang 
Currently, we only take local machine as our execution environment. So, all the data can be from disk.
Let's make the whole training process work first.  Python code is involved in the runtime breaks our design. But it only happens in feeding data or passing data to our training process. Most other training logic is fine with compile-time design.

Reply:
@QiJune sure, thanks for explaining! Could you create an issue for it and put into the TODO in the Github project?

Reply:
@helinwang Yes, I have created a issue #4613. But we have not a Project now for supporting distributed training feature. Let's put it later.